File: Server/src/main/java/com/mercury/zippit/net/codec/handshake/HandshakeResponseEncoder.java
package com.mercury.zippit.net.codec.handshake;

import com.mercury.zippit.configuration.Version;
import com.mercury.zippit.utilities.ByteBufUtilities;
import io.netty.buffer.ByteBuf;
import io.netty.channel.ChannelHandlerContext;
import io.netty.handler.codec.MessageToByteEncoder;

import static com.mercury.zippit.net.codec.handshake.HandshakeConstants.INVALID_SERVICE;
import static com.mercury.zippit.net.codec.handshake.HandshakeConstants.OUTDATED;

/**
 * @author Harrison, Alias: Hc747, Contact: harrisoncole05@gmail.com
 * @version 1.0
 * @since 27/10/17
 */
public final class HandshakeResponseEncoder extends MessageToByteEncoder<HandshakeRequest> {

	private final Version version;

	public HandshakeResponseEncoder(Version version) {
		super(HandshakeRequest.class);
		this.version = version;
	}

	@Override
	protected void encode(ChannelHandlerContext context, HandshakeRequest request, ByteBuf response) {
		if (!version.equals(request.getVersion())) {
			fail(response, OUTDATED);
			return;
		}

		HandshakeRequestEndpoint endpoint = request.getEndpoint();

		if (endpoint == null) {
			fail(response, INVALID_SERVICE);
			return;
		}

		response.writeBoolean(true);
		endpoint.handle(context);
	}

	private void fail(ByteBuf buffer, String reason) {
		buffer.writeBoolean(false);
		ByteBufUtilities.writeString(buffer, reason);
	}

}

File: Server/src/main/java/com/mercury/zippit/net/codec/handshake/HandshakeResponseEncoder.kt
package com.mercury.zippit.net.codec.handshake

import com.mercury.zippit.configuration.Version
import com.mercury.zippit.net.codec.handshake.HandshakeConstants.FAILURE
import com.mercury.zippit.net.codec.handshake.HandshakeConstants.INVALID_SERVICE
import com.mercury.zippit.net.codec.handshake.HandshakeConstants.OUTDATED
import com.mercury.zippit.utilities.ByteBufUtilities.writeString
import io.netty.buffer.ByteBuf
import io.netty.channel.ChannelHandlerContext
import io.netty.handler.codec.MessageToByteEncoder
import java.util.logging.Logger

/**
 * @author Harrison, Alias: Hc747, Contact: harrisoncole05@gmail.com
 * @version 1.0
 * @since 30/10/17
 */
class HandshakeResponseEncoder(private val version: Version) : MessageToByteEncoder<HandshakeRequest>(HandshakeRequest::class.java) {

    companion object {

        private val logger = Logger.getLogger(HandshakeResponseEncoder::class.java.simpleName)

    }

    override fun encode(context: ChannelHandlerContext, request: HandshakeRequest, response: ByteBuf) {

        logger.info("Processing request: $request at: ${System.currentTimeMillis()}")

        if (version != request.version) {
            fail(response, OUTDATED)
            return
        }

        if (request.endpoint == null) {
            fail(response, INVALID_SERVICE)
            return
        }

        request.endpoint.handle(context)
        response.writeByte(request.endpoint.ordinal)
    }

    private fun fail(buffer: ByteBuf, reason: String) {
        //TODO:
        buffer.writeByte(FAILURE)
        writeString(buffer, reason)
    }

}